File: magnet-market/python/utils/decay_matrix.py
"""
utils/decay_matrix.py
Implements the Time-Decayed Memory Matrix for IPDA (Institutional Price Discovery Addresses).
Computes rolling exponential-weighted price references over configurable windows.

Core Formula:
$$
\text{IPDA}_k(t) = \sum_{i=t-L_k}^{t} P_i \cdot e^{-\lambda(t-i)}
$$
"""

import numpy as np
from numpy.lib.stride_tricks import sliding_window_view
from typing import List, Optional
import pandas as pd


def compute_decay_matrix(
    prices: np.ndarray,
    windows: List[int] = [20, 40, 60],
    decay_rate: float = 0.15
) -> np.ndarray:
    """
    Computes the decayed memory matrix for specified lookback windows.
    
    Parameters
    ----------
    prices : np.ndarray
        1D array of closing prices or OHLCV median.
    windows : List[int]
        List of IPDA memory depths $L_k$ (e.g., [20, 40, 60]).
    decay_rate : float
        Lambda $\\lambda$ controlling time-decay speed. Higher values forget older memory faster.
        
    Returns
    -------
    np.ndarray
        Matrix of shape (N, K) where N = len(prices), K = len(windows).
        Columns correspond to each window in `windows`. Initial NaNs pad incomplete windows.
    """
    prices = np.asarray(prices, dtype=np.float64)
    N = len(prices)
    K = len(windows)
    result = np.full((N, K), np.nan)

    for k, L in enumerate(windows):
        if L > N:
            continue
            
        # Weight vector: [exp(-lambda*(L-1)), ..., exp(0)]
        # Oldest bar gets strongest decay, current bar gets weight = 1
        weights = np.exp(-decay_rate * np.arange(L)[::-1])
        
        # Create overlapping windows without copying memory
        windows_view = sliding_window_view(prices, L)
        
        # Vectorized dot product: each row dotted with decay weights
        result[L-1:, k] = np.dot(windows_view, weights)
        
    return result


def compute_ipda_extremums(
    ipda_matrix: np.ndarray,
    method: str = "local"
) -> np.ndarray:
    """
    Identifies windowed extremum references from the decayed IPDA matrix.
    
    Parameters
    ----------
    ipda_matrix : np.ndarray
        Output from compute_decay_matrix.
    method : str
        'local' for immediate memory peaks, 'global' for session-wide bounds.
        
    Returns
    -------
    np.ndarray
        2D array of (support, resistance) levels per timestep.
    """
    # Ignore NaN rows
    valid_rows = ~np.isnan(ipda_matrix).any(axis=1)
    supports = np.where(valid_rows, np.min(ipda_matrix[:, 1:], axis=1), np.nan)
    resistances = np.where(valid_rows, np.max(ipda_matrix[:, 1:], axis=1), np.nan)
    return np.column_stack((supports, resistances))

File: magnet-market/rust/utils/decay_matrix.rs
// utils/decay_matrix.rs
// Implements the Time-Decayed Memory Matrix for IPDA (Institutional Price Discovery Addresses).
// Computes rolling exponential-weighted price references over configurable windows.

// Core Formula:
// $$
// \text{IPDA}_k(t) = \sum_{i=t-L_k}^{t} P_i \cdot e^{-\lambda(t-i)}
// $$

use ndarray::{Array1, Array2, ArrayView1, Axis};
use ndarray::s;
use std::f64;

fn compute_decay_matrix(
    prices: &Array1<f64>,
    windows: Vec<usize>,
    decay_rate: f64,
) -> Array2<f64> {
    let n = prices.len();
    let k = windows.len();
    let mut result = Array2::<f64>::zeros((n, k));

    for (i, &l) in windows.iter().enumerate() {
        if l > n {
            continue;
        }

        // Weight vector: [exp(-lambda*(L-1)), ..., exp(0)]
        let weights: Array1<f64> = Array1::from_iter((0..l).rev().map(|x| (-decay_rate * x as f64).exp()));

        // Create overlapping windows
        let windows_view = prices.windows(l);

        // Vectorized dot product: each row dotted with decay weights
        for (j, window) in windows_view.enumerate() {
            result[(j + l - 1, i)] = window.dot(&weights);
        }
    }
    result
}

fn compute_ipda_extremums(
    ipda_matrix: &Array2<f64>,
    method: &str,
) -> Array2<f64> {
    let valid_rows = ipda_matrix.axis_iter(Axis(0)).map(|row| !row.iter().any(|&x| x.is_nan())).collect::<Vec<_>>();
    let supports = valid_rows.iter().enumerate().map(|(i, &valid)| {
        if valid {
            ipda_matrix.slice(s![i, 1..]).iter().cloned().fold(f64::INFINITY, f64::min)
        } else {
            f64::NAN
        }
    }).collect::<Array1<f64>>();

    let resistances = valid_rows.iter().enumerate().map(|(i, &valid)| {
        if valid {
            ipda_matrix.slice(s![i, 1..]).iter().cloned().fold(f64::NEG_INFINITY, f64::max)
        } else {
            f64::NAN
        }
    }).collect::<Array1<f64>>();

    Array2::from_shape_vec((supports.len(), 2), supports.iter().chain(resistances.iter()).cloned().collect()).unwrap()
}
